app/post.go

```diff
@@ -1174,3 +1174,131 @@ func (a *App) MaxPostSize() int {
 
 	return maxPostSize
 }
+
+func (a *App) countMentionsFromPost(user *model.User, post *model.Post) (int, *model.AppError) {
+	channel, err := a.GetChannel(post.ChannelId)
+	if err != nil {
+		return 0, err
+	}
+
+	if channel.Type == model.CHANNEL_DIRECT {
+		// Return -1 to indicate that the mention count should be equal to the post count
+		return -1, nil
+	}
+
+	channelMember, err := a.GetChannelMember(channel.Id, user.Id)
+	if err != nil {
+		return 0, err
+	}
+
+	keywords := addMentionKeywordsForUser(
+		map[string][]string{},
+		user,
+		channelMember.NotifyProps,
+		&model.Status{Status: model.STATUS_ONLINE}, // Assume the user is online since they would've triggered this
+		true, // Assume channel mentions are always allowed for simplicity
+	)
+	commentMentions := user.NotifyProps[model.COMMENTS_NOTIFY_PROP]
+	checkForCommentMentions := commentMentions == model.COMMENTS_NOTIFY_ROOT || commentMentions == model.COMMENTS_NOTIFY_ANY
+
+	// A mapping of thread root IDs to whether or not a post in that thread mentions the user
+	mentionedByThread := make(map[string]bool)
+
+	isCommentMention := func(post *model.Post, otherPosts map[string]*model.Post) bool {
+		if post.RootId == "" {
+			// Not a comment
+			return false
+		}
+
+		if mentioned, ok := mentionedByThread[post.RootId]; ok {
+			// We've already figured out if the user was mentioned by this thread
+			return mentioned
+		}
+
+		// Whether or not the user was mentioned because they started the thread
+		mentioned := otherPosts[post.RootId].UserId == user.Id
+
+		// Or because they commented on it before this post
+		if !mentioned && commentMentions == model.COMMENTS_NOTIFY_ANY {
+			for _, otherPost := range otherPosts {
+				if otherPost.Id == post.Id {
+					continue
+				}
+
+				if otherPost.RootId != post.RootId {
```

Comment:
if post and otherposts are from the same thread, Root shouldn't be the same always except for the initial one? wouldnt this be easier like `otherPost.RootId == ""`? but at the same time, this wouldn't happen if post and otherPost were both the initial ones? Isn't this condition way too complicated with many different cases, maybe we should be more explicit on what are we trying to achieve or at least comment it so people is aware of all the posibilities in the condition?

Reply:
wouldn't this miss the case where otherPost.RootId == post.Id, I mean, when the post is the root of the other post?

Reply:
That case gets covered by [this line](https://github.com/mattermost/mattermost-server/pull/11966/files#diff-317b12c12f71e311f13813107f4d8550R1219). That case is shared between the "Any" and "Root" options for comment mentions, so I didn't repeat it here